calyx-py/calyx/py_ast.py

```diff
@@ -13,6 +20,77 @@ def emit(self):
         print(self.doc())
 
 
+class FileTable:
+    # global counter to ensure unique ids
+    counter: int = 0
+    table: Dict[str, int] = {}
+
+    @staticmethod
+    def get_fileid(filename):
+        if filename not in FileTable.table:
+            FileTable.table[filename] = FileTable.counter
+            FileTable.counter += 1
+        return FileTable.table[filename]
+
+    @staticmethod
+    def emit_fileinfo_metadata():
+        contents = []
+        for filename, fileid in FileTable.table.items():
+            contents.append(f"{fileid}: {filename}")
+        return block("FILES", contents, with_curly=False)
+
+
+class PosTable:
+    counter: int = 0
+    table: Dict[(int, int), int] = {}  # (fileid, linenum) -> positionId
```

Comment:
nit: typing annotation should be
```suggestion
    table: Dict[Tuple[int, int], int] = {}  # (fileid, linenum) -> positionId
```